### DISCUSSION TITLE
How to check for dayjs type with zod?

### DISCUSSION THREAD

**Discussion by anthonyma94** [colinhacks/zod#1259] [Q&A] [answered]
I'm using [Day.js](https://day.js.org) extensively in my application. The way to check for dayjs type is to check instanceof dayjs (the function):
```js
import dayjs from "dayjs";
const date = dayjs();
console.log(date instanceof dayjs) // true
```

The param type of `z.instanceof` only accepts classes with constructors, but it works in runtime for checking instanceof dayjs. This means that I have to typecast dayjs as any every time I need it in instanceof:
```ts
const dateType = z.instanceof(dayjs as any);
```

`z.infer` also infers this as any, which doesn't really help me.

How can I infer dayjs properly?

**Comment by dvakatsiienko** [comment, marked as answer]
The solution didn't worked for me causing an error.

```
Cannot read properties of undefined (reading 'name')
TypeError: Cannot read properties of undefined (reading 'name') at Object.instanceOfType [as instanceof] (webpack-internal:///./node_modules/zod/lib/index.mjs:3065:43)
```

But this works:
```ts
import dayjs, { type Dayjs } from 'dayjs';

...

z.instanceof(dayjs as unknown as typeof Dayjs)
```

**Reply by jhigdon** [reply]
thank you, I needed this as well, though I am unsure why :)

**Reply by aki-caffeine** [reply]
This answer is very helpful to me. However, I don't know why this coding works properly.

Is there anyone who can explain?

**Reply by almogzur** [reply]
Thank You

**Comment by infodusha** [comment]
If anyone want to haver a proper version with no type-cast:

```ts
const zodDay = z.custom<Dayjs>((val) => val instanceof dayjs, 'Invalid date');
```

**Reply by Zain-Bin-Arshad** [reply]
Or you can use `isDayjs` like this:

```
const zodDay = z.custom<Dayjs>((val) => dayjs.isDayjs(val), "Invalid date");

**Reply by jeremy-code** [reply]
IMO this is best, since z.instanceof is itself a wrapper over z.custom (see [colinhacks/zod/blob/main/src/types.ts#L5357](https://github.com/colinhacks/zod/blob/main/src/types.ts#L5357). I've experimented with a custom class that extends `ZodType`, but it seems unnecessary especially since `z.date()` already has validations.

For a slightly more useful error message, check out the following:
```ts
import dayjs, { type Dayjs } from "dayjs";
import { z } from "zod";

export const DayjsSchema = z.custom<Dayjs>(
  (data) => dayjs.isDayjs(data),
  (input) => ({
    // if not interested in error map, just return a string
    // message: `Expected Dayjs, received ${z.getParsedType(input)}`,
    params: { expected: "Dayjs", received: z.getParsedType(input) },
  }),
);

const errorMap: z.ZodErrorMap = (issue, ctx) => {
  if (issue.code === z.ZodIssueCode.custom) {
    if (issue.params?.expected && issue.params?.received) {
      return {
        message: `Expected ${issue.params.expected}, received ${issue.params.received}`,
      };
    }
  }
  return { message: ctx.defaultError };
};
z.setErrorMap(errorMap);

DayjsSchema.parse(dayjs());
DayjsSchema.parse(dayjs); // Expected Dayjs, received function
DayjsSchema.parse(new Date()); // Expected Dayjs, received date
DayjsSchema.parse(undefined); // Expected Dayjs, received undefined
```

**Comment by lidak-patrik** [comment]
I know this is old but for anybody doing this, i would recoment adding `.isValid()` function of dayjs.

I have used something like this with MUI DatePicker
```
export const dayjsDateSchema = zod.custom<Dayjs>((val) => dayjs(val).isValid() || val === null, {
  message: 'Invalid date',
});
```

I have done this because the dayjs variable can have an invalid value, for example day/month/year can be null, and the `isDayjs()` function cannot recognize this.

**Reply by amechain-dtml** [reply]
and what did you use as a type in an interface for dayjs?

**Reply by lidak-patrik** [reply]
If you mean the type that is casted into the generic `<Dayjs>` that is a type/interface directly from the dayjs package.
`import type { Dayjs } from 'dayjs';`
Alternativaly you should be able to use <any/unknown> and check if its dayjs format with the `.isDayjs()` function and then check with `.isValid()`.
